walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,82 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<String> relevantAttendees = new ArrayList<>();
+    request.getAttendees().forEach(attendee -> relevantAttendees.add(attendee));
+
+    List<TimeRange> affectedTimeRanges = new ArrayList<>(); 
+    // get all affected time ranges
+    for (Event event : events) {
+      Set<String> scheduledAttendees = event.getAttendees(); 
+      for (String attendee : relevantAttendees) {
+        if (scheduledAttendees.contains(attendee)) {
+          // this event affects the final result
+          affectedTimeRanges.add(event.getWhen());
+          break;
+        }
+      }
+    }
+
+    long minRequiredDuration = request.getDuration();
+    long maxDuration = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true).duration();
+
+    List<TimeRange> availableRanges = new ArrayList<>();
+    if (minRequiredDuration > maxDuration) {
+      return availableRanges;
+    } else if (affectedTimeRanges.size() == 0) {
+      availableRanges.add(TimeRange.WHOLE_DAY);
+      return availableRanges;
+    }
+
+    // sort by start time
+    Collections.sort(affectedTimeRanges, TimeRange.ORDER_BY_START);
+
+    // consider the possibility of a range that starts at START
+    TimeRange firstAffectedRange = affectedTimeRanges.get(0); // no error because there is at least one time range
```

Comment:
Can we merge START and END case into the general case?

Reply:
I could merge START into the general case! But I would need to create and insert a "dummy" `TimeRange` into `affectedTimeRanges`. But I don't see how the END case could be merged. :/

Reply:
I might be wrong but another approach to this problem might eliminate the START and END case. For example, right now we're tackling the problem starting with the first affectedTimeRange. What if we start with the first available range? E.g, the first empty range would be `START_OF_DAY - start_of_first_affected_range`. You can still keep track of the endsLastSoFar, and based on that you can find the next available range. And the last range will be `end_of_lendsLastSoFar - END_OF_DATY`, this way they can all be handled in the same way without creating special handling for any cases.

Reply:
Hmm... But in the case of `end_of_lendsLastSoFar - END_OF_DAY`, I will still have to check whether it is the last time range. If otherwise, I don't see how I can determine that the end point should be `END_OF_DAY`.

Reply:
Having to check whether or not it's the last time range isn't something that seems quite hard to me, e.g, you can check via index. My point about not handling start and end separately is that, with special handling on start and end case, we have three entries for adding an available time range. If the definition of available time range changes, e.g, one day we want to know avaible time range only within SIN time (imagine finding availabel time slots for attendees worldwide to attend an event at SG office), we'll only need to update the logic in once place, without copying the code three times for each handling. 

Just some personal thought about flexibility of implementation, I won't block you on this. Currently implementation works well enough tho!

Reply:
Yup, I understand where you are coming from! (: Thus, I have reduced the number of cases, which I need to consider, to two. 😅 As always, thanks for the feedback!